File: queue/queue.py
class Node:
  def __init__(self, value):
    self.value = value
    self.next = None

class Queue:
  
  def __init__(self, head_value):
    self.head = Node(head_value)
    self.tail = None

  def enqueue(self, value):
    new_node = Node(value)
    if not self.tail:
      self.head.next = new_node
      self.tail = new_node
    else:
      self.tail.next = new_node
      self.tail = new_node

  def dequeue(self):
    current = self.head
    self.head = current.next
    current.next = None
    return current.value

  def peek(self):
    return self.head.value

queue = Queue(1)
queue.enqueue(2)
queue.enqueue(3)
queue.enqueue(4)
queue.enqueue(5)
queue.enqueue(6)
queue.enqueue(7)

print(queue.head.value)
print(queue.head.next.value)
print(queue.head.next.next.value)
print(queue.head.next.next.next.value)
print(queue.head.next.next.next.next.value)
print(queue.head.next.next.next.next.next.value)
print(queue.head.next.next.next.next.next.next.value)

print(queue.dequeue())
print(queue.dequeue())
print(queue.dequeue())
print(queue.dequeue())
print(queue.dequeue())
print(queue.dequeue())

print(queue.head.value, 'new head value')

File: queue/queue.ts
type TypeNode<T> = {
  value: T
  next?: TypeNode<T> 
  prev?: TypeNode<T>
} 

class Queue<T> {
  public length: number = 0
  public head: TypeNode<T>
  public tail: TypeNode<T> | null

  constructor(public headValue: T){
    this.head = { value: headValue } as TypeNode<T>
    this.tail = null
  }

  enqueue(item: T) {
    const newNode = { value: item } as TypeNode<T>
    if (!this.tail){
      this.head.next = newNode
      this.tail = newNode
    } else {
      this.tail.next = newNode
      this.tail = newNode
      this.length++
    }
  }

  deque():T {
    const current = this.head
    if (this.head.next){
      this.head = this.head.next
    }
    this.length--
    return current.value
  } 

  peek():T {
    console.log(this.head.value)
    return this.head.value
  }
}

const queue = new Queue<number>(1)

queue.enqueue(3)
queue.enqueue(9)
queue.enqueue(2)
queue.enqueue(8)

console.log(queue.head.value)
console.log(queue.head.next?.value)
console.log(queue.head.next?.next?.value)
console.log(queue.head.next?.next?.next?.value)
console.log(queue.head.next?.next?.next?.next?.value)
console.log(queue.tail?.value)

console.log(queue.deque())
console.log(queue.deque())
console.log(queue.deque())
// console.log(queue.deque())

console.log(queue.head.next)